_posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md

```diff
@@ -0,0 +1,207 @@
+---
+permalink: "/{{ year }}/{{ month }}/{{ day }}/multi-store-custom-telemetry-with-shared-data"
+title: "multi-store - Custom Telemetry with shared data"
+published_date: "2019-01-18 12:00:00 +0100"
+layout: post.liquid
+data:
+  route: blog
+---
+
+Last year I implemented a new feature for Firefox Telemetry that changes how we can collect and analyze data with different requirements in regard to user privacy & frequency.
+This post will shine some light on the (rather simple) implementation and usage.
+
+### Intro: What is Firefox Telemetry?
+
+In order to understand how Firefox performs in the wild, it can collect a bunch of performance metrics and other information.
+The [Telemetry component](https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/) provides APIs used by the various other parts to gather this data
+and is responsible for storing, collecting and sending this data in what we call "pings", a periodic collection of measurements.
+Telemetry data is only ever send out if the user agreed to it (see "Data Collection and Use" in the "Privacy & Security" preferences of your Firefox).
+
+Data can be collected in one of 3 different formats: histograms, scalars & events (see [our collection overview](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/index.html)).
+Firefox sends this data in the ["main" ping](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/main-ping.html) once in a while (usually roughly daily) and clears out the stored data locally.
+A "main" ping always corresponds to a subsession, which itself is part of a session, this is further explained [in our Session concepts](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/concepts/sessions.html).
+
+People working with the data can therefore make certain assumptions on how to interpret the data from multiple pings across sessions and subsessions.
+
+### The problem
+
+Some data should not be correlated with other data due to privacy concerns.
+So far we had to push Telemetry users to create custom pings and keep track of their own data.
+If they rely on scalars or histograms as recorded my Telemetry core, but send a custom ping in different intervals, they can't make valid assumptions about the metrics, as they might have been reset in between.
+This leads to weird hacks or unnecessary code duplication. Additionally we can't provide any help or support for custom data and our tools can't handle it automatically (e.g. to generate dashboards).
+
+### A solution
+
+**Multi-Store**.
+
+Every probe was always tied to the schedule of the "main" ping (read: subsession/session).
```

Comment:
Are we calling them probes or metrics? I think we settled on "metrics", at least on mobile.